lib/instrumentation/modules/cassandra-driver.js

```diff
@@ -0,0 +1,186 @@
+      const span = agent.buildSpan()
+      if (!span) {
+        return original.apply(this, arguments)
+      }
+
+      const queryStrings = queries.map(toQueryString)
+      const querySummaries = queryStrings.map(sqlSummary)
+      const query = queryStrings.join(';\n')
+      const querySummary = querySummaries.join('; ')
+
+      span.type = 'db.cassandra.query'
+      span.setDbContext({ statement: query, type: 'cassandra' })
+      span.name = `BATCH ${querySummary}`
+      span.start()
+
+      function resolve () {
+        span.end()
+      }
+
+      // Wrap the callback
+      const index = arguments.length - 1
+      const cb = arguments[index]
+      const isPromise = typeof cb !== 'function'
+      if (!isPromise) {
+        arguments[index] = function wrappedCallback () {
+          resolve()
+          return cb.apply(this, arguments)
+        }
+      }
+
+      const ret = original.apply(this, arguments)
+
+      if (isPromise) {
+        if (typeof ret.then === 'function') {
+          ret.then(resolve, resolve)
+        } else {
+          agent.logger.error('unable to identify span exit point for cassandra-driver')
+        }
+      }
+
+      return ret
+    }
+  }
+
+  function wrapExecute (original) {
+    return function wrappedExecute (query, params, options, callback) {
+      const span = agent.buildSpan()
+      if (!span) {
+        return original.apply(this, arguments)
+      }
+
+      span.type = 'db.cassandra.query'
+      span.setDbContext({ statement: query, type: 'cassandra' })
+      span.name = sqlSummary(query)
+      span.start()
+
+      function resolve () {
+        span.end()
+      }
+
+      // Wrap the callback
+      const index = arguments.length - 1
+      const cb = arguments[index]
+      const isPromise = typeof cb !== 'function'
+      if (!isPromise) {
+        arguments[index] = function wrappedCallback () {
+          resolve()
+          return cb.apply(this, arguments)
+        }
+      }
+
+      const ret = original.apply(this, arguments)
+
+      if (isPromise) {
+        if (typeof ret.then === 'function') {
+          ret.then(resolve, resolve)
+        } else {
+          agent.logger.error('unable to identify span exit point for cassandra-driver')
+        }
+      }
+
+      return ret
+    }
+  }
+
+  function wrapEachRow (original) {
+    return function wrappedEachRow (query, params, options, rowCallback, callback) {
+      const span = agent.buildSpan()
+      if (!span) {
+        return original.apply(this, arguments)
+      }
+
+      span.type = 'db.cassandra.query'
+      span.setDbContext({ statement: query, type: 'cassandra' })
+      span.name = sqlSummary(query)
+      span.start()
+
+      // Wrap the callback
+      const index = arguments.length - 1
+      const hasRowCallback = typeof arguments[index - 1] === 'function'
+
+      function resolve () {
+        span.end()
+      }
+
+      if (hasRowCallback) {
+        const cb = arguments[index]
```

Comment:
It's technically possible for the user to call the batch function with a proper rowCallback in the 2nd to last position and a non-function argument in the last position. I was wondering if we could deal with that in some way... what do you think?

Reply:
What would we do in that situation? It's an invalid input, and I'm not sure how we could recover from that other than just leaving the span un-ended. 🤔

Reply:
No we can't recover, and in we probably need a general way to "destroy" a span if we find that we can't detect when it ends. But adding that ability to spans is probably out of scope for this PR.

In regards to invalid input in general, I try to just forward that invalid input to the module being patched so that it can use its normal error handling.

Reply:
fixed